### PR TITLE
Add Feature to test Unity Packages

### LINKED ISSUE #71: Feature: Tooling to run tests for Unity packages
Hi!

First of all, thanks to you and all contributors for this repository ❤️ 

# Background

I'm currently open sourcing some Unity packages (tools etc.) from our company to be used with the new Unity Package Manager.
I'm using this repository's action to run tests for those packages.

- Package I'm currently developing: https://github.com/3d-group/unity-simple-notifications (open source)
- Unity docs on custom packages: https://docs.unity3d.com/Manual/CustomPackages.html

# Problem

It seems like Unity requires an Unity project to run tests.

To circumvent this problem I created an empty Unity project at my repository and ignored it with `.npmignore` file and just ran tests there using the projectPath variable.

But I cannot help thinking that it is a bit silly to include empty Unity project just for running tests 😄

# Suggested solution

I think this could be automated and included in this action like so:

1. Check if there is package.json in project root in order to detect if project is an upm (unity package manager) package.
    or perhaps this could be told via env variable?
2. Create new Unity project (either copy it from somewhere or download from Unity website?)
3. Edit manifest.json at UnityProject/Packages/

Add following lines to `dependencies` and `testables` sections:
```json
{
  "dependencies": {
    "com.company.package": "https://github.com/company/repository.git"
  },
  "testables": [ "com.company.package" ]
}
```

Perhaps Unity project could also be cached for future action runs?

### PR DESCRIPTION
**Pull request by trudeaua21**
#### Changes
- closes #71 
- Adds a new argument to the action named `packageMode`, which is `false` by default. If `true`, then the action tests a Unity Package rather than a Unity Project
- In package mode, all functionality is the same for testing packages as for testing project, except for 
  - caching, which isn't available
  - an explicit Unity version is required
  -  in package mode, the CLI utility `jq` is used, which means that any docker image used to run the action in package mode must be compatible with `apt-get`
  -  tests are run by creating an empty Unity project at runtime, adding the package to that project, and testing the project.
- A test Unity package has been added to the repo, as well as several workflow items in `main.yml` to test the action's usage on the package.

##### Areas for Improvement
Here's some clear areas for improvement with the feature that I haven't yet implemented. If needed, I will try implementing these before the PR is merged, but if not, then I'll fill out a new issue for these improvements.
- Caching may be able to be implemented in package mode by caching the empty project that gets created every time the tests run. The project would need to be cached _before_ the package is added to its manifest and testables.
- Having a unity version of `auto` shouldn't be too hard to implement - I'm fairly sure it can just be taken from the package's `package.json`. The main reason it's not part of this PR is just because I was focusing on getting eyes on the MVP first.

#### Checklist

- [x] Read the contribution [guide](../CONTRIBUTING.md) and accept the [code](../CODE_OF_CONDUCT.md) of conduct
- [x] Readme (updated or not needed)
- [x] Tests (added, updated or not needed)

### CODE REVIEW

**Comment by GabLeRoux** [comment]
Nice one! Thanks for the contribution :D

> in package mode, the CLI utility jq is used, which means that any docker image used to run the action in package mode must be compatible with apt-get

We now have support for windows based docker images so another area of improvement might be to add a condition to get this working for windows based images as well, but that can definitely be in a next step. 👍

**Review by davidmfinol — APPROVED** [review]
LGTM. Thanks!

File: dist/entrypoint.sh [hunk 1]
```diff
@@ -7,6 +7,26 @@
...
+
+  # install jq
+  apt-get update \
+    &&  apt-get upgrade -y --force-yes \
+    &&  apt-get install -y --force-yes \
+        jq \
+    &&  apt-get clean \
+    &&  rm -rf /var/cache/apt/archives/* /var/lib/apt/lists/*
+fi
+
```
Comment (webbertakken) [inline dist/entrypoint.sh:29]: While this isn't wrong per se, installing JQ during CI runtime is probably not a good idea. 

I would suggest that we install JQ on the images instead, so that we can simply use it.

Also note that the `.sh` files should be as thin as possible, as the logic is all contained in the javascript part of the image (where possible).

File: dist/steps/run_tests.sh [hunk 2]
```diff
@@ -7,6 +7,65 @@
...
+  fi
+
+  PACKAGE_NAME=$(cat "$PACKAGE_JSON_PATH" | jq -r ".name")
+
+  if [ -z $PACKAGE_NAME ]; then 
+    echo "Unable to parse package name from package.json. Aborting..."
+    exit 1
+  fi
+
+  echo "Package name found: ${PACKAGE_NAME}"
```
Comment (webbertakken) [inline dist/steps/run_tests.sh:39]: Ideally these checks would move to the javascript side. That way 
- we don't have to pull 5GB if we know it's going to fail in the first place
- it will be easier to continue with our efforts to consolidate core logic from test-runner and builder
- we don't need to install jq on runtime or in the images.

File: dist/steps/run_tests.sh [hunk 3]
```diff
@@ -7,6 +7,65 @@
...
+
+  echo "Creating an empty Unity project to add the package to."
+
+  unity-editor \
+    -batchmode \
+    -createProject "./TempProject" \
```
Comment (webbertakken) [inline dist/steps/run_tests.sh:45]: Since `TempProject` is used in multiple places, it might make sense to add it to a descriptive env variable (like `EMPTY_PROJECT` or `PACKAGE_TESTING_PROJECT`).

Which would also make this line self documenting, where:
```bash
  UNITY_PROJECT_PATH="./TempProject"

  # becomes

  UNITY_PROJECT_PATH="$PACKAGE_TESTING_PROJECT"
```

**Review by webbertakken — CHANGES_REQUESTED** [review]
I really like your solution. Amazing work!

A few comments about code structure and semantics, before we merge this.

File: dist/entrypoint.sh (hunk 1 above)
Comment (trudeaua21) [reply dist/entrypoint.sh:29]: Makes sense! 

I'm not super well versed in Docker yet, so just to clarify - are you suggesting I modify the `Dockerfile` to install `jq` there, or modify the [`unityci/editor` image](https://hub.docker.com/r/unityci/editor) to add a `jq` installation? 

I believe you're talking about modifying the image itself - in that case, is [this repo](https://github.com/game-ci/docker) the place to do so?

Thanks for all the feedback, I really appreciate it! 😄

Comment (trudeaua21) [reply dist/entrypoint.sh:29]: Ah, only just now saw that your feedback made it so we don't need to install `jq` - my bad! I'll get to that right away! 😄

File: dist/steps/run_tests.sh (hunk 2 above)
Comment (trudeaua21) [reply dist/steps/run_tests.sh:39]: I agree with moving these checks over to the javascript side - however:

> we don't need to install jq on runtime or in the images.

I don't think this is necessarily true. We still need `jq` for the operation of adding the package name to the generated project's manifest and testables:

```
  PACKAGE_MANIFEST_JSON=$(cat "$PACKAGE_MANIFEST_PATH")
  echo "$PACKAGE_MANIFEST_JSON" | \
    jq \
    --arg packageName "$PACKAGE_NAME" \
    --arg projectPath "$UNITY_PROJECT_PATH" \
    '.dependencies += {"\($packageName)": "file:\($projectPath)"} | . += {testables: ["\($packageName)"]}' \
    > "$PACKAGE_MANIFEST_PATH"

  UNITY_PROJECT_PATH="./TempProject"
```

Granted, we don't really _need_ `jq` per se, but it just saves us from having to manually find and insert the necessary strings into the manifest file. Let me know if you'd prefer that route over using `jq` and I'll work at making that change!

**Comment by trudeaua21** [comment]
By the way, I noticed the pre-commit hook wasn't working for me a few commits back. I had just been manually doing the pre-commit steps on my own for a little while, but I decided to dig a bit deeper and found that the problem was just that the file for the pre-commit hook wasn't executable. I updated the file to be executable and the hook started working for me again, so I figured I was good to commit that change - if that's not right feel free to call it out!

File: dist/steps/run_tests.sh (hunk 2 above)
Comment (webbertakken) [reply dist/steps/run_tests.sh]: I didn't say that we don't need `jq`. The operative word here was meant to be `install during runtime`.

Javascript has npm packages to simplify working with JSON. Though that said javascript natively understands JSON, so you can just [parse](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse) it.

**Comment by webbertakken** [comment]
> By the way, I noticed the pre-commit hook wasn't working 
> ...
> I updated the file to be executable and the hook started working for me again, so I figured I was good to commit that change 

Great, I've seen the hooks not working for more people indeed. Thank you very much. We should probably apply the same fix to all our repos.

File: dist/steps/run_tests.sh (hunk 2 above)
Comment (trudeaua21) [reply dist/steps/run_tests.sh]: Makes sense, sorry for the confusion!

Just to clarify, would I install `jq` in the `Dockerfile`, or would we install it in [the Docker images themselves](https://github.com/game-ci/docker)

Thank you!

Comment (webbertakken) [reply dist/steps/run_tests.sh]: Indeed, in the docker images themselves.

https://packages.ubuntu.com/bionic/jq 

The size seems to be very small so it's ok to add it to  the base image (`images/ubuntu/base`)

**Comment by andywiecko** [comment]
Hey, great work!

I just want to share with you my idea.
I apologize if I should post this as a separate issue, but I thought it could be easily implemented in the following PR if my thinking makes any sense.

---

Currently, custom Unity Packages do not support the "git dependencies" (and I read somewhere at unity forum that it will probably never be supported). In the case when one has a package that depends on other custom packages, the current solution is no longer valid, unless packages are available in some registry.
Consider the example `package.json`, where "packageA" and "packageB" are custom unity packages:

```json
{
    "name": "package",
    "version": "0.1.0",
    
    ...

    "dependencies": {
        "packageA": "1.1.0",
        "packageB": "1.2.0"
    }
}
```

Then one has to add them as a dependency at project `manifest.json`:

```json
{
  "dependencies": {
     "packageA": "git+https://github.com/user/packageA#1.1.0",
     "packageB": "git+https://github.com/user/packageB#1.2.0",
     
    ...
  }
}
```

I propose to add a parameter, like `additionalDependencies`, then  the job configuration for my example can look like following

```yaml
   // ...

    - uses: game-ci/unity-test-runner
        id: tests
        with:
          projectPath: ${{ matrix.projectPath }}
          unityVersion: ${{ matrix.unityVersion }}
          testMode: ${{ matrix.testMode }}
          artifactsPath: ${{ matrix.testMode }}-artifacts
          packageMode: true
          additionalDependencies:
             - git+https://github.com/user/packageA#1.1.0
             - git+https://github.com/user/packageB#1.2.0
             
   // ...
```

Best,
Andrzej

**Comment by JesseTG** [comment]
Thankfully with the prevalence and simplicity of OpenUPM there's a lot less need for `git`-specified dependencies.

**Comment by trudeaua21** [comment]
Honestly, in the current state of this work, I don't think it's even possible to test packages that depend on any external registry. I would hope that the solution is as simple as just tossing the scoped registries' information into the manifest of the generated Unity project we use for testing manually (similarly to how adding the package under test to the manifest manually currently):

```
{
   // HOPEFULLY WE COULD JUST ADD THE REGISTRY INFO HERE
    "scopedRegistries": [
        {
            "name": "General",
            "url": "https://example.com/registry",
            "scopes": [
                "com.example", "com.example.tools.physics"
            ]
        },
        {
            "name": "Tools",
            "url": "https://mycompany.example.com/tools-registry",
            "scopes": [
                "com.example.mycompany.tools"
            ]
        }
    ],
    "dependencies": {
        "com.sample.project": "location"
    }
}
```

I would guess that packages that depend on other packages in the official Unity registry will work fine with the runner in its current state (since I'm assuming that Unity's package manager will be able to resolve them without issue), but I will need to check that to make sure. 

**TL;DR** packages that depend on external registries or other package dependency approaches probably don't work with the current package support of the runner from this PR. I think it's definitely feasible to add support for those.

I think I would need to ask the maintainers their preference here, since there's two options:
- After I finish troubleshooting the issues from those who tested this, we can merge this work with the functionality as-is, with an explanation in the documentation that packages with dependencies that _are not_ from the Unity package registry are not supported
- We keep this same PR open, and wait to merge it until after support for the packages described above are added

**Comment by webbertakken** [comment]
> I think I would need to ask the maintainers their preference here, since there's two options

We mostly prefer small iterations where possible so if this is a chunk that works for your use case already and it doesn't create forward incompatibility we're good to merge it.

**Comment by JesseTG** [comment]
Is there any reason this PR wouldn't work with third-party (but still NPM-powered) registries? I'd really like to use it for a test pipeline for a package I'm working on, but I can only do that if it can load dependencies from OpenUPM.

**Comment by trudeaua21** [comment]
> Is there any reason this PR wouldn't work with third-party (but still NPM-powered) scoped registries? I'd really like to use it for a test pipeline for a package I'm working on, but I can only do that if it can load dependencies from OpenUPM.

@JesseTG There isn't a reason past that it will take more time to implement and then test. It seems like it would be simple to implement support for scoped registries, based on the fact that the package manifest for Unity Projects has scoped registry URLs present, indicating that everything Unity needs for scoped registries can be found in the package manifest.

_That said_, I believed this body of work would be simple as well, and while all things considered, it kind of was, it still has taken a good few months due to some unforseen complications 😅 

I'll probably work on it after this gets merged, but the functionality could still take some time.

**Comment by saszer** [comment]
Can this also be extended with the feature to do builds with just packages as well in future?

Ps, being eyeing this, hope it gets merged soon. Extra features with package dependency compatibilities etc, can be a separate issue imo.

**Comment by andywiecko** [comment]
@saszer Good idea for the builds! I reported this in unity.builder repo: https://github.com/game-ci/unity-builder/issues/363

**Comment by JesseTG** [comment]
> > Is there any reason this PR wouldn't work with third-party (but still NPM-powered) scoped registries? I'd really like to use it for a test pipeline for a package I'm working on, but I can only do that if it can load dependencies from OpenUPM.
> 
> @JesseTG There isn't a reason past that it will take more time to implement and then test. It seems like it would be simple to implement support for scoped registries, based on the fact that the package manifest for Unity Projects has scoped registry URLs present, indicating that everything Unity needs for scoped registries can be found in the package manifest.
> 
> _That said_, I believed this body of work would be simple as well, and while all things considered, it kind of was, it still has taken a good few months due to some unforseen complications 😅
> 
> I'll probably work on it after this gets merged, but the functionality could still take some time.

Ah, I think I see the problem. `package.json` (the file you're extracting info from) doesn't generally care about particular registries, but `manifest.json` (the file you're creating) does. If scoped registries aren't available in either file, you'd need a third source.

But I think it would be easiest to require that packages tested with this action define the necessary scoped registries in `package.json`. Then you could just copy the `scopedRegistries` object and insert it directly into the generated `manifest.json`. What do you think?

**Comment by trudeaua21** [comment]
> But I think it would be easiest to require that packages tested with this action define the necessary scoped registries in `package.json`. Then you could just copy the `scopedRegistries` object and insert it directly into the generated `manifest.json`. What do you think?

I actually didn't know that scoped registries could be defined in the `package.json` of a Unity package - do you have an example of this? Pretty much all I know about `package.json` structure just comes from this page (and the related docs), and I can't find an example of having scoped registries in `package.json`: https://docs.unity3d.com/Manual/upm-manifestPkg.html

Not to say I don't believe it's possible (lord knows there's some fun undocumented stuff with Unity packages), I just couldn't find it at a moment's glance. If it's possible, that would make implementation of this feature easier 😄

**Comment by JesseTG** [comment]
> > But I think it would be easiest to require that packages tested with this action define the necessary scoped registries in `package.json`. Then you could just copy the `scopedRegistries` object and insert it directly into the generated `manifest.json`. What do you think?
> 
> I actually didn't know that scoped registries could be defined in the `package.json` of a Unity package - do you have an example of this? Pretty much all I know about `package.json` structure just comes from this page (and the related docs), and I can't find an example of having scoped registries in `package.json`: https://docs.unity3d.com/Manual/upm-manifestPkg.html
> 
> Not to say I don't believe it's possible (lord knows there's some fun undocumented stuff with Unity packages), I just couldn't find it at a moment's glance. If it's possible, that would make implementation of this feature easier 😄

Here's the unabridged `manifest.json` I'm using to develop a package I'm working on. However, this manifest will not be included in the repository.

```json
{
  "scopedRegistries": [
    {
      "name": "package.openupm.com",
      "url": "https://package.openupm.com",
      "scopes": [
        "com.cysharp.unitask"
      ]
    }
  ],
  "dependencies": {
    "com.cysharp.unitask": "2.2.5",
    "com.unity.ide.rider": "3.0.12",
    "com.unity.ide.visualstudio": "2.0.14",
    "com.unity.ide.vscode": "1.2.5",
    "com.unity.nuget.newtonsoft-json": "2.0.0",
    "com.unity.test-framework": "1.1.31",
    "com.unity.modules.animation": "1.0.0",
    "com.unity.modules.assetbundle": "1.0.0",
    "com.unity.modules.audio": "1.0.0",
    "com.unity.modules.director": "1.0.0",
    "com.unity.modules.imageconversion": "1.0.0",
    "com.unity.modules.imgui": "1.0.0",
    "com.unity.modules.jsonserialize": "1.0.0",
    "com.unity.modules.screencapture": "1.0.0",
    "com.unity.modules.ui": "1.0.0",
    "com.unity.modules.uielements": "1.0.0",
    "com.unity.modules.unityanalytics": "1.0.0",
    "com.unity.modules.unitywebrequest": "1.0.0",
    "com.unity.modules.unitywebrequestassetbundle": "1.0.0",
    "com.unity.modules.unitywebrequestaudio": "1.0.0",
    "com.unity.modules.unitywebrequesttexture": "1.0.0",
    "com.unity.modules.unitywebrequestwww": "1.0.0"
  }
}
```

**Comment by trudeaua21** [comment]
Yeah, that seems like it would be feasible to implement in that way. 

Just an update for anyone following the thread - here's what's going to happen before this gets merged:
1. My contribution which adds `jq` to the installed tools of the `game-ci` docker images, which the `unity-test-runner` leverages, is pretty much complete (hopefully - there's some failing checks but that could be a fluke, since they didn't fail on a different run of the test workflow). This will need to get merged before this feature is functional again.
2. After that gets merged, I'm going to run a test to confirm that this feature works on a basic, barebones package (go to the issue for this feature if you'd like to see a package someone put together which will help test this)
3. After that, I'm going to add some tests to my own package that reference a dependency from the Unity package registry to confirm that Unity package registry dependencies work with this feature. If this fails, I'll have to re-evaluate completion time.

After that, I feel like the first iteration of this feature is pretty much done! 😄 

After this feature gets merged, provided someone else doesn't do it first, I'll make an issue or two to track the improvements to this feature (Windows image support, support for scoped registries, support for git dependencies, etc.)

Thanks to everyone for your help and patience 😄

**Comment by trudeaua21** [comment]
[Per my comment on the issue](https://github.com/game-ci/unity-test-runner/issues/71#issuecomment-1103147012), next monday I'll have the time to start working on this again! 

It looks like there's [another PR](https://github.com/game-ci/unity-test-runner/pull/182) with some changes that will likely get merged before I finish this up, so I'll make sure to update this to ensure compatibility with the features from that PR.

**Comment by trudeaua21** [comment]
Alrighty, I feel like the first iteration of this feature is pretty much done! I've gone ahead and re-requested review.

[Here's an example of the feature working on an existing package.](https://github.com/trudeaua21/EasyAccessibilityPackage/actions/runs/2243067904)

I'm not sure why the workflow is failing - I'm a bit stumped because [the problem isn't happening on my fork](https://github.com/trudeaua21/unity-test-runner/actions/runs/2230587793).

My steps after this gets merged will be as follows:

1. Document possible extensions of this work/limitations of this work in a new issue or two.
2. [Add some more in depth documentation for this new feature](https://github.com/game-ci/documentation)

Thanks for your patience everyone! 😄

**Comment by trudeaua21** [comment]
Actually, in case this does get reviewed, please wait to merge (assuming everything ends up looking good) until I implement the fix I've described [on the issue](https://github.com/game-ci/unity-test-runner/issues/71#issuecomment-1115743825) for packages located on the repository root folder

File: action.yml [hunk 4]
```diff
@@ -5,14 +5,14 @@ inputs:
   unityVersion:
     required: false
     default: 'auto'
-    description: 'Version of unity to use for testing the project. Use "auto" to get from your ProjectSettings/ProjectVersion.txt'
+    description: 'Version of unity to use for testing the project. Use "auto" to get from your ProjectSettings/ProjectVersion.txt. ⚠️ If testing a Unity Package, this field is required and cannot be set to "auto".'
```
Comment (webbertakken) [inline action.yml:8]: Is there a way to detect that a Unity Package is tested instead of the regular case?

Perhaps we can just smoothen this over inside the auto strategy?

Perhaps something like this?
```ts
  if (input.packageMode) return
```

File: dist/steps/run_tests.sh [hunk 5]
```diff
@@ -7,6 +7,69 @@
 UNITY_PROJECT_PATH="$GITHUB_WORKSPACE/$PROJECT_PATH"
 echo "Using project path \"$UNITY_PROJECT_PATH\"."
 
+#
+# Create an empty project for testing if in package mode
+#
```
Comment (webbertakken) [inline dist/steps/run_tests.sh:12]: Should this whole block be moved after all the general settings?

Right before the line that says `# Overall info`

**Review by webbertakken — APPROVED** [review]

**Comment by trudeaua21** [comment]
BTW, I've been pretty busy the past couple weeks, and the rest of this week - not sure when I'll get around to addressing the feedback. 

Also, there's going to be one more functionality tweak I'd like to add before this gets released - currently, the action fails if the directory to the package is the root directory of the repository. I've confirmed a workaround exists by making a copy of the root directory to some subdirectory before the tests run, but I need to add that to automatically happen within the action.

**Comment by webbertakken** [comment]
> BTW, I've been pretty busy the past couple weeks, and the rest of this week - not sure when I'll get around to addressing the feedback.

No rush :)

> Also, there's going to be one more functionality tweak I'd like to add before this gets released - currently, the action fails if the directory to the package is the root directory of the repository. I've confirmed a workaround exists by making a copy of the root directory to some subdirectory before the tests run, but I need to add that to automatically happen within the action.

Generally we support having projects in the root. It might be a bad idea to just copy the root directory, especially with regards to project size and file permissions. Do you know why it's failing exactly, when the project is in the root?

**Comment by trudeaua21** [comment]
Hey all - my day job has been pretty busy the last couple months and left me little time to dedicate to this. However, this Wednesday/Thursday I'll finally have some time to dig back into this and _hopefully_ address all the feedback 🤞

File: dist/steps/run_tests.sh (hunk 5 above)
Comment (trudeaua21) [reply dist/steps/run_tests.sh:41]: Makes sense; I moved it!

File: action.yml (hunk 4 above)
Comment (trudeaua21) [reply action.yml:8]: I added a thrown error for this case on [line 109 of `input.ts`](https://github.com/game-ci/unity-test-runner/pull/164/files#diff-4d2973eda0a779e451cdd0304253311207a08a08e31ef58809cc738edd23d14dR108-R117):

```
    if (packageMode) {
      if (unityVersion === 'auto') {
        throw new Error(
          'Package Mode is enabled, but unityVersion is set to "auto". unityVersion must manually be set in Package Mode.',
        );
      }

      packageName = this.getPackageNameFromPackageJson(projectPath);
      this.verifyTestsFolderIsPresent(projectPath);
    }
```

Does this address the problem or no? Not totally sure if this is what you were going for. 

I'd like to add an "auto" strategy for package mode, but not until after the base feature is released.

**Comment by trudeaua21** [comment]
> Generally we support having projects in the root. It might be a bad idea to just copy the root directory, especially with regards to project size and file permissions. Do you know why it's failing exactly, when the project is in the root?

Spent a good chunk of the day looking through the logs and trying out different things in an attempt to figure this out. I've eliminated a few possible conditions that I initially suspected were the problem, including:
- presence of `.git`, `.github`, and `_activate-license` files
- Permissions of the folder containing the package being tested or any of its files (not including files in any subdirectories)
- Permissions of the folder of the generated project or any of its files (not including files in any subdirectories)

The specific thing going wrong is that when the **package being tested is the root of the repository** , the test command fails. Roughly 20,000 lines of logs accompany the failure, but before those lines start this gets logged:

```
2022-07-01T01:50:11.0455269Z ###########################
2022-07-01T01:50:11.0455805Z #   Testing in playmode  #
2022-07-01T01:50:11.0456324Z ###########################
2022-07-01T01:50:11.0456608Z 
2022-07-01T01:51:06.3636835Z 
2022-07-01T01:51:06.3637523Z Aborting batchmode due to failure:
2022-07-01T01:51:06.3638242Z Scripts have compiler errors.
```

Another oddity I found - when the package folder **IS** the directory root and the action fails, there's many statements in the logs like these:
```
2022-07-01T01:51:06.9336866Z GUID [a6351153cf5e1af468cae5dbc72d1a04] for asset 'Packages/com.trudeaua.easyaccessibility/TempProject/Library/PackageCache/com.unity.testtools.codecoverage@1.1.1/CHANGELOG.md.meta' conflicts with:
2022-07-01T01:51:06.9337280Z   'Packages/com.unity.testtools.codecoverage/CHANGELOG.md.meta'
2022-07-01T01:51:06.9337483Z Assigning a new guid.
```

None of these GUID conflicts are present in situations where the package folder **is not** the directory root, including cases where whatever workflow is calling the action moves the package folder out of the directory root. 

Here's logs for a run of the action where everything works as intended: 
[1_tests.txt](https://github.com/game-ci/unity-test-runner/files/9024623/1_tests.txt)


Here's logs for a run of the action where the action fails due to the package folder being the directory root: 
[1_tests.txt](https://github.com/game-ci/unity-test-runner/files/9024630/1_tests.txt)

**Comment by trudeaua21** [comment]
Hey, so update on this. 

I never was able to figure out why there was a problem with packages in root. I was able to verify that the problem could be worked around by moving the project out of root in a GitHub Workflow step before the test runner runs.

I've got to be completely honest here - given some recent decisions Unity has made as a company, I really can't justify continuing work on their platform right now to myself. In addition, those same decisions have completely taken the wind out of my sails on this work as well. 

Aside from digging deeper into the issue with the package being in root (I think that debugging that problem would take **a large chunk** of my time, which I don't think would be manageable given my lack of motivation on Unity work right now), I'd be cool with doing whatever is needed to sort of wrap up work on this (like writing documentation, opening further issues, etc). 

I would also understand if the maintainers would rather not merge this feature as it is right now, as it might not really be considered "done" as it would require a workaround for some users. 

More than anything, I'd like to really say thank you to the maintainers for your help and attention in developing this feature. You all rock and it's been a pleasure working with you on this. 

I'm sorry if this makes any trouble for you all.

**Comment by trudeaua21** [comment]
Sounds great, thanks for your understanding! Sometime in the next week or two I'll re-test that the package works for my initial use-case and add some documentation for the new features (and for the workaround for having packages in root).

**Comment by trudeaua21** [comment]
Hey all, see https://github.com/game-ci/unity-test-runner/issues/71#issuecomment-1502141778 for a quick update on this.

Essentially, I never really got around to wrapping this ticket up, but I'm planning on chipping away at it over the next few weekends to try to get it wrapped up.

Apologies that this hung around so long - hopefully I can get this off your plates!

**Comment by trudeaua21** [comment]
@davidmfinol @webbertakken Sorry for the ping, but I need your input on the questions at the bottom.

This is just about ready for review now - I've caught it up with the latest changes and updated the documentation in a couple of spots, as well as adding a check in the script which throws an error if it can't find the `jq` command (which could happen on a custom Docker container).

**Here are the list of caveats/limitations of the feature as it stands right now:**
- The test runner will fail if the directory of the package being tested is the same as
  the root directory of the repository. 
- The test runner can only test packages on Linux runners - Windows runners are currently not
  supported.
- Packages with dependencies outside of the Unity Registry (for example, packages with Git Dependencies, Scoped registries, or having anything to do with UPM) have not been tested, so the test runner may or may not currently work with such packages.
- There is currently no caching set up for package mode.
- A `unityVersion` configuration of 'auto' cannot be used when testing a Unity package. The Unity
  version must be manually set in the action.
- If using the `customImage` parameter to use a custom Docker image to test the package, that image
  must have [`jq`](https://jqlang.github.io/jq/) installed, or else the test runner will error. It
  is used to add the package being tested to a temporary Unity project's package manifest at
  runtime.

**I just have a couple questions that I need answered in order to finish up now:**
1. For all except the last item (regarding jq on custom images) of the list of limitations above, I plan on opening issues to address those work items as feature requests/bug reports. Should I put them all into one issue, or each item in its own issue?
2. I've forked the [documentation repo](https://github.com/game-ci/documentation) to add Package Mode documentation. My plan is to first, create the issues as stated in question 1. Then, I will re-request review for this PR, and request review for my documentation PR in tandem. Does this workflow sound fine? (Note: the documentation PR will require the issues to be opened first, as I plan to link to them within a "limitations" section of the documentation for package mode).
3. For some reason, when coming back to this, the precommit hasn't been playing nicely with my default git config regarding line endings. I've tweaked my git config to try to adjust, but that somehow ended up in CRLF line endings making their way into one of the bash scripts, causing it to fail. To remedy this, I just ran `dos2unix` on the whole repo - however, this has resulted in ~16 extra changed files on this PR. Each of those ~16 files seem to only have CRLF -> LF changes, and all the ones I've seen are only on specific lines. **My question is** - would you like me to try to revert my branch back a couple commits and deal with the line endings problem I faced in a more focused way than just running `dos2unix`, or is it fine to leave the branch as-is?

Thank you so much to the maintainers. Again, I apologize this has taken so long, and I really appreciate your patience and assistance throughout this whole process.

**Comment by webbertakken** [comment]
Just heading out but I can answer quickly now:

1. Prefer separate issues where possible
2. Sounds fine to me
3. If you're using auto-crlf you shouldn't have this problem. I'll list my config below. Ideally we wouldn't have 16 files change that aren't related, but if it really doesn't work then it's no disaster.

```ini
# `~/.gitconfig

# ...

[core]
eol = native
autocrlf = input

# ...
```

**Comment by trudeaua21** [comment]
Alright - I've finished up all the documenting that I can think of, and filled out a PR for documentation of this feature in the [documentation repo](https://github.com/game-ci/documentation/pull/398). 

I'll need some input on that PR from the maintainers - I left it in draft status since I couldn't get the documentation repo to build locally for some reason, but I've made all the changes I intended to, so it's only in draft status out of formality given the contribution guidelines. 

**But great news - this PR is ready for review again!** 🚀 

@davidmfinol Tagging you for visibility since I wasn't able to re-request your review - sorry if that wasn't necessary!

File: unity-package-with-correct-tests/com.fake.notarealpackage/Tests/Runtime/fake.notarealpackage.RuntimeTests.asmdef [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+    "name": "fake.notarealpackage.RuntimeTests",
```
Comment (webbertakken) [inline unity-package-with-correct-tests/com.fake.notarealpackage/Tests/Runtime/fake.notarealpackage.RuntimeTests.asmdef:2]: Prefer `example.com` or `com.example` instead of using actual website names.

That domain is meant for such uses specifically.

**Review by webbertakken — APPROVED** [review]

File: unity-package-with-correct-tests/com.fake.notarealpackage/Tests/Runtime/fake.notarealpackage.RuntimeTests.asmdef (hunk 6 above)
Comment (trudeaua21) [reply unity-package-with-correct-tests/com.fake.notarealpackage/Tests/Runtime/fake.notarealpackage.RuntimeTests.asmdef]: Good call! I've addressed this in my most recent commit - the package is now called `com.example.testpackage`.

It has ~40 changes, but that's mainly just because I renamed the folder for the package to the new package name. The key changes are:

- Renamed the package folder itself
- Updated the package.json with the new package name & updating the info in it to better match the new name
- Updated all the assembly definitions to match the new name
- Updated the `main.yml` tests to point at the new package name

If that was the only note, then I have nothing else to add 👍

Comment (webbertakken) [reply unity-package-with-correct-tests/com.fake.notarealpackage/Tests/Runtime/fake.notarealpackage.RuntimeTests.asmdef]: Cool. Looks good to me. 

Probably good if some of @davidmfinol, @AndrewKahr, @frostebite, @GabLeRoux, @lazerwalker could have a look. And then hopefully we can release it soon :)

Comment (cloudymax) [reply unity-package-with-correct-tests/com.fake.notarealpackage/Tests/Runtime/fake.notarealpackage.RuntimeTests.asmdef]: What a journey! I just caught up with the history on this one. I don't see any major issues with the limitations, `jq` is a pretty common requirement, using only Linux runners for the time-being seems fine for a initial release of a new feature.

Seems like there have already been several users who have used the fork successfully with all issues reported already having been addressed.

Great work!

**Review by AndrewKahr — APPROVED** [review]
This LGTM!

Really really amazing work, I know this feature has been talked about a lot and so finally getting support for it will be a huge step forward!

Again awesome work!

**Review by davidmfinol — APPROVED** [review]

**Comment by webbertakken** [comment]
Released in https://github.com/game-ci/unity-test-runner/releases/tag/v2.2.0

**Comment by timcassell** [comment]
> I've got to be completely honest here - given some recent decisions Unity has made as a company, I really can't justify continuing work on their platform right now to myself. In addition, those same decisions have completely taken the wind out of my sails on this work as well.

This was a year ago, but Unity's latest stunt almost makes it seem like you saw the future!
